es-app/src/guis/GuiInfoPopup.cpp

```diff
@@ -0,0 +1,114 @@
+#include "guis/GuiInfoPopup.h"
+#include "Renderer.h"
+#include "components/TextComponent.h"
+#include "Log.h"
+
+GuiInfoPopup::GuiInfoPopup(Window* window, std::string message, int duration) :
+	GuiComponent(window), mMessage(message), mDuration(duration), running(true)
+{
+	mFrame = new NinePatchComponent(window);
+	float maxWidth = Renderer::getScreenWidth() * 0.9f;
+	float maxHeight = Renderer::getScreenHeight() * 0.2f;
+
+	std::shared_ptr<TextComponent> s = std::make_shared<TextComponent>(mWindow,
+		"",
+		Font::get(FONT_SIZE_MINI),
+		0x444444FF,
+		ALIGN_CENTER);
+
+	// we do this to force the text container to resize and return an actual expected popup size
+	s->setSize(0,0);
+	s->setText(message);
+	mSize = s->getSize();
+
+	// confirm the size isn't larger than the screen width, otherwise cap it
+	if (mSize.x() > maxWidth) {
+		s->setSize(maxWidth, mSize[1]);
+		mSize[0] = maxWidth;
+	}
+	if (mSize.y() > maxHeight) {
+		s->setSize(mSize[0], maxHeight);
+		mSize[1] = maxHeight;
+	}
+
+	// add a padding to the box
+	int paddingX = Renderer::getScreenWidth() * 0.03f;
+	int paddingY = Renderer::getScreenHeight() * 0.02f;
+	mSize[0] = mSize.x() + paddingX;
+	mSize[1] = mSize.y() + paddingY;
+
+	float posX = Renderer::getScreenWidth()*0.5f - mSize.x()*0.5f;
+	float posY = Renderer::getScreenHeight() * 0.02f;
+
+	setPosition(posX, posY, 0);
+
+	mFrame->setImagePath(":/frame.png");
+	mFrame->fitTo(mSize, Eigen::Vector3f::Zero(), Eigen::Vector2f(-32, -32));
+	addChild(mFrame);
+
+	// we only init the actual time when we first start to render
+	mStartTime = 0;
+
+	mGrid = new ComponentGrid(window, Eigen::Vector2i(1, 3));
+	mGrid->setSize(mSize);
+	mGrid->setEntry(s, Eigen::Vector2i(0, 1), false, true);
+	addChild(mGrid);
+}
+
+GuiInfoPopup::~GuiInfoPopup()
+{
+
+}
+
+void GuiInfoPopup::render(const Eigen::Affine3f& parentTrans)
+{
+	// we use identity as we want to render on a specific window position, not on the view
+	Eigen::Affine3f trans = getTransform() * Eigen::Affine3f::Identity();
+	if(running && updateState())
+	{
+		// if we're still supposed to be rendering it
+		Renderer::setMatrix(trans);
+		renderChildren(trans);
+	}
+}
+
+bool GuiInfoPopup::updateState()
```

Comment:
I know :( 

I struggled with it for a while, and in the end I didn't get the NinePatchComponent to fade in and out using Lambda. It's probably because it's using the image Alpha, rather than using the Opacity setting - and I imagine that the LambdaAnimation is using Opacity.

That's why - if you read the updateState method - I need to change the Opacity for one of them, and the color alpha for the other one.

It's terrible.

I could consider redefining the SetOpacity method for the NinePatchComponent and see if that would solve that.